### PR TITLE
panstarrs_get_lightcurve using lsdb

### PR DESCRIPTION
**Pull request by jkrick**
having some trouble catching exceptions for FileNotFound in the call to lsdb's crossmatch with the Panstarrs catalog.  It would be nice if we could somehow catch this error so that the crossmatch doesn't fail if a file is missing.  Instead, it would be ideal if one of the objects fails the crossmatch, that the crossmatch still finishes with the other objects complete, and the final result just looks like no match was found for that trouble-causing object/file.  

Helpdesk request submitted to MAST 9/19 about the missing file.

### CODE REVIEW

**Comment by jkrick** [comment]
MAST helpdesk responded that they will fix the file problem on S3 before Oct 7, so this PR will wait for that fix to move forward, so that testing can be completed.

**Comment by jkrick** [comment]
Now fully functional!
This PR includes:
- new panstarrs function which uses lsdb to access a hipscat parquet in S3
- removes the 'old' functions which use the panstarrs API
- changes name in plotting to be in line with actual name of the dataset "Pan-STARRS" (but is too hard for me to type every time)
- added runtime info to the intro cell (800s to run the full notebook) @bsipocz is this too long?

Notes
- along the way I discovered missing files in the panstarrs S3 parquets, MAST fixed those, and Troy assures me that he has checks for these things when he pushes data to S3.  So I think that is one problem that got fixed, that I hope we will not have to see in the future.
- This function is not actually faster for the sample size of 30 that we use in the notebook, but will be faster than the API for larger samples, and in general accessing data in parquet format in the cloud is the direction we want to be going.
- this might be the first calls to lsdb in our notebooks?  but probably not the last.

This will close #165 
This uses help from: https://github.com/astronomy-commons/lsdb/issues/416

**Comment by bsipocz** [comment]
> 800s to run the full notebook

It's a bit long, but for fornax it's totally acceptable unless we hit a VM limit. We currently skip this notebook due to the ZTF private bucket access.

(and I don't know why we see the error with the classifier notebook, that should not be affected by this PR)

File: light_curves/requirements_light_curve_generator.txt [hunk 1]
```diff
@@ -14,6 +14,9 @@ astroquery>=0.4.8.dev0
 acstools
 lightkurve
 alerce
+lsdb
+s3fs
```
Comment (bsipocz) [inline light_curves/requirements_light_curve_generator.txt:18]: We don't directly use s3fs so don't need to add it to this file.

Also do we have to manually install it as it's an undeclared dependency of lsdb? (would be nice to be able to pick it up automatically though an extras, although it's a theoretical ask for the lsdb developers as here we already install it as part of `pandas[aws]`

File: light_curves/requirements_light_curve_generator.txt [hunk 2]
```diff
@@ -14,6 +14,9 @@ astroquery>=0.4.8.dev0
...
+nested_pandas
```
Comment (bsipocz) [inline light_curves/requirements_light_curve_generator.txt:19]: no need to declare it as we don't directly use it (and lsdb should install it if it requires it)

File: light_curves/light_curve_generator.md [hunk 3]
```diff
@@ -131,9 +135,9 @@ get_yang_sample(coords, labels)   #2018ApJ...862..109Y
 # a balance between speed of running the light curves and whatever
 # the ML algorithms would like to have
 
-# num_normal_QSO = 5000
-# zmin, zmax = 0, 10
-# randomize_z = False
+#num_normal_QSO = 30
+#zmin, zmax = 0, 10
+#randomize_z = False
```
Comment (bsipocz) [inline light_curves/light_curve_generator.md:140]: are these changes intended?

File: light_curves/light_curve_generator.md [hunk 4]
```diff
@@ -84,6 +86,8 @@ import astropy.units as u
 import pandas as pd
 from astropy.table import Table
 
+import numpy as np
+
```
Comment (bsipocz) [inline light_curves/light_curve_generator.md:90]: This is not used anywhere
```suggestion
```

File: light_curves/code_src/panstarrs_functions.py [hunk 5]
```diff
@@ -1,236 +1,16 @@
...
-    
-    return(dresults)
+from data_structures import MultiIndexDFObject
 
 
-#Do a panstarrs search
-def panstarrs_get_lightcurves(sample_table, *, radius=1/3600):
-    """Searches panstarrs for light curves from a list of input coordinates.  This is the MAIN function.
+#panstarrs light curves from hipscat catalog in S3 using lsdb
+def panstarrs_get_lightcurves(sample_table, radius):
```
Comment (bsipocz) [inline light_curves/code_src/panstarrs_functions.py:12]: Please keep the API: radius should have a sensible default and then being an optional but mandatorily keyword argument. Or at least keep it as a mandatory keyword argument. (e.g. I want to avoid having functions that end up with a call signature of positional arguments like this `functionname(a, b, None, False, True, True)`)

File: light_curves/light_curve_generator.md [hunk 6]
```diff
@@ -250,19 +258,35 @@ print('WISE search took:', time.time() - WISEstarttime, 's')
...
-The function to retrieve lightcurves from Pan-STARRS currently uses their API; based on this [example](https://ps1images.stsci.edu/ps1_dr2_api.html).  This search is not efficient at scale and we expect it to be replaced in the future.
+The function to retrieve lightcurves from Pan-STARRS uses a version of both the object and light curve catalogs that are stored in the cloud and accessd using [lsdb](https://docs.lsdb.io/en/stable/).  This function is efficient at large scale (sample sizes > ~1000).
 
 ```{code-cell} ipython3
 panstarrsstarttime = time.time()
 
-panstarrs_search_radius = 1.0/3600.0    # search radius = 1 arcsec
+panstarrs_search_radius = 1.0 # search radius in arcsec
 # get panstarrs light curves
-df_lc_panstarrs = panstarrs_get_lightcurves(sample_table, radius=panstarrs_search_radius)
```
Comment (bsipocz) [inline light_curves/light_curve_generator.md:260]: keep this one. Being explicit, especially in a demo, is important.

File: light_curves/light_curve_generator.md [hunk 7]
```diff
@@ -362,7 +386,7 @@ n_workers = 8
 heasarc_kwargs = dict(catalog_error_radii={"FERMIGTRIG": "1.0", "SAXGRBMGRB": "3.0"})
 ztf_kwargs = dict(nworkers=None)  # the internal parallelization is incompatible with Pool
 wise_kwargs = dict(radius=1.0, bandlist=['W1', 'W2'])
-panstarrs_kwargs = dict(radius=1.0/3600.0)
+panstarrs_search_radius = 1.0 # arcsec
```
Comment (bsipocz) [inline light_curves/light_curve_generator.md:389]: nitpick, but I think it would ultimately be nice if these radiuses could be any astropy quantity rather than a float, of which the user should keep track if arcsec, or arcmin, or degrees, depending on which datasets they are querying. This comment likely will go beyond this PR into a follow-up

**Review by bsipocz — COMMENTED** [review]
Some comments. Mostly minor, but I picked up on some things to make the API nicer (e.g. quantity inputs, or being explicit with parameters)

File: light_curves/light_curve_generator.md (hunk 3 above)
Comment (jkrick) [reply light_curves/light_curve_generator.md:140]: From my perspective, it doesn't matter what the values of these are since they are commented out.  I do change them when I am working on the code and testing different sample sizes, and then most often forget what they originally were and leave them at whatever the last value I tried.  Does that cause a problem?  Or is it the spaces between the # and the commands that is concerning?

File: light_curves/requirements_light_curve_generator.txt (hunk 1 above)
Comment (jkrick) [reply light_curves/requirements_light_curve_generator.txt:18]: I removed s3fs from the requirements file.  I don't quite understand what the ask is for lsdb, so please open an issue over there if it is desirable.

File: light_curves/code_src/panstarrs_functions.py (hunk 5 above)
Comment (jkrick) [reply light_curves/code_src/panstarrs_functions.py]: done

File: light_curves/light_curve_generator.md (hunk 4 above)
Comment (jkrick) [reply light_curves/light_curve_generator.md]: done

File: light_curves/light_curve_generator.md (hunk 6 above)
Comment (jkrick) [reply light_curves/light_curve_generator.md:260]: done

File: light_curves/light_curve_generator.md (hunk 3 above)
Comment (bsipocz) [reply light_curves/light_curve_generator.md:140]: It's the adding unrelated and unnecessary changes. I see why it happens, but it would be nice to figure out a convenient way to change the workflow so only intended and necessary changes get added. 
I added this topic to my list for the ipac visit.

File: light_curves/light_curve_generator.md (hunk 7 above)
Comment (jkrick) [reply light_curves/light_curve_generator.md:387]: I turned this into a new issue to track it, but not necessarily add it to this PR.

File: light_curves/requirements_light_curve_generator.txt (hunk 2 above)
Comment (jkrick) [reply light_curves/requirements_light_curve_generator.txt]: I removed nested_pandas from the requirements file.

**Review by jkrick — COMMENTED** [review]
Thanks for looking it over and the comments, I think I got them all.

File: light_curves/light_curve_generator.md (hunk 3 above)
Comment (jkrick) [reply light_curves/light_curve_generator.md:138]: yes!  I don't know how to do this, so perfect to discuss later this month.

File: light_curves/requirements_light_curve_generator.txt [hunk 8]
```diff
@@ -14,6 +14,9 @@ astroquery>=0.4.8.dev0
 acstools
 lightkurve
 alerce
+lsdb
+#s3fs
+#nested_pandas
```
Comment (bsipocz) [inline light_curves/requirements_light_curve_generator.txt:19]: I'm sorry for nitpicking, but these need to be removed and not commented out. Rule of thumb, only leave commented out lines when they serve a clear purpose or a message to the user. Leaving in stuff that are not needed any more doesn't fall in that category.
```suggestion
lsdb
dask[distributed]
```

File: light_curves/light_curve_generator.md [hunk 9]
```diff
@@ -250,19 +256,35 @@ print('WISE search took:', time.time() - WISEstarttime, 's')
...
+The function to retrieve lightcurves from Pan-STARRS uses a version of both the object and light curve catalogs that are stored in the cloud and accessd using [lsdb](https://docs.lsdb.io/en/stable/).  This function is efficient at large scale (sample sizes > ~1000).
 
 ```{code-cell} ipython3
 panstarrsstarttime = time.time()
 
-panstarrs_search_radius = 1.0/3600.0    # search radius = 1 arcsec
+panstarrs_search_radius = 1.0 # search radius = 1 arcsec
 # get panstarrs light curves
-df_lc_panstarrs = panstarrs_get_lightcurves(sample_table, radius=panstarrs_search_radius)
+df_lc_panstarrs = panstarrs_get_lightcurves(sample_table, radius = panstarrs_search_radius)
```
Comment (bsipocz) [inline light_curves/light_curve_generator.md:266]:
```suggestion
df_lc_panstarrs = panstarrs_get_lightcurves(sample_table, radius=panstarrs_search_radius)
```

File: light_curves/light_curve_generator.md [hunk 10]
```diff
@@ -131,9 +133,9 @@ get_yang_sample(coords, labels)   #2018ApJ...862..109Y
 # a balance between speed of running the light curves and whatever
 # the ML algorithms would like to have
 
-# num_normal_QSO = 5000
-# zmin, zmax = 0, 10
-# randomize_z = False
+#num_normal_QSO = 30
+#zmin, zmax = 0, 10
+#randomize_z = False
```
Comment (bsipocz) [inline light_curves/light_curve_generator.md:138]:
```suggestion
# num_normal_QSO = 5000
# zmin, zmax = 0, 10
# randomize_z = False
```

File: light_curves/light_curve_generator.md [hunk 11]
```diff
@@ -70,9 +72,9 @@ MAST, HEASARC, & IRSA Fornax teams
 
 This cell will install them if needed:
 
-```{code-cell} ipython3#
+```{code-cell} ipython3
 # Uncomment the next line to install dependencies if needed.
-# !pip install -r requirements_light_curve_generator.txt
+!pip install -r requirements_light_curve_generator.txt
```
Comment (bsipocz) [inline light_curves/light_curve_generator.md:77]: Don't commit changes to this line.
```suggestion
# !pip install -r requirements_light_curve_generator.txt
```

File: light_curves/code_src/panstarrs_functions.py [hunk 12]
```diff
@@ -1,236 +1,16 @@
...
-    return(dresults)
+from data_structures import MultiIndexDFObject
 
 
-#Do a panstarrs search
-def panstarrs_get_lightcurves(sample_table, *, radius=1/3600):
-    """Searches panstarrs for light curves from a list of input coordinates.  This is the MAIN function.
+#panstarrs light curves from hipscat catalog in S3 using lsdb
+def panstarrs_get_lightcurves(sample_table, radius = 1):
+    """Searches panstarrs hipscat files for light curves from a list of input coordinates.
```
Comment (bsipocz) [inline light_curves/code_src/panstarrs_functions.py:13]:
```suggestion
# panstarrs light curves from hipscat catalog in S3 using lsdb
def panstarrs_get_lightcurves(sample_table, *, radius=1):
    """Searches panstarrs hipscat files for light curves from a list of input coordinates.  
```

File: light_curves/code_src/panstarrs_functions.py [hunk 13]
```diff
@@ -1,236 +1,16 @@
...
-    
-    """
-    dconstraints = {'objID': objid}
-    dcolumns = ("""objID,detectID,filterID,obsTime,ra,dec,psfFlux,psfFluxErr,psfMajorFWHM,psfMinorFWHM,
-            psfQfPerfect,apFlux,apFluxErr,infoFlag,infoFlag2,infoFlag3""").split(',')
-    # strip blanks and weed out blank and commented-out values
-    dcolumns = [x.strip() for x in dcolumns]
-    dcolumns = [x for x in dcolumns if x and not x.startswith('#')]
+import lsdb
+from dask.distributed import Client
```
Comment (bsipocz) [inline light_curves/code_src/panstarrs_functions.py:5]: dask is directly used, and thus has to be added to the requirements file.

File: light_curves/requirements_light_curve_generator.txt (hunk 1 above)
Comment (bsipocz) [reply light_curves/requirements_light_curve_generator.txt]: xref: https://github.com/astronomy-commons/lsdb/issues/436

**Review by bsipocz — APPROVED** [review]
cleanup changes to be committed and then this is good to go.

File: light_curves/light_curve_generator.md (hunk 7 above)
Comment (troyraen) [reply light_curves/light_curve_generator.md:387]: > it would ultimately be nice if these radiuses could be any astropy quantity

Generally, I agree, but in this case (as I recall) the scale_up.md notebook needs these to be floats (or ints) because the functions are called from the command line. So if they get changed to astropy quantities we will need to implement a workaround in scale_up.md.

OTOH, we'll probably want to revisit the parallel cell in this notebook because this PR removes Pan-STARRS from the `multiprocessing` call so now it has to wait for the others to finish and then it runs separately. It would be nice to get them all to actually run in parallel again. Potentially, we can come up with a solution that actually works at scale, and then we wouldn't need scale_up.md anymore. The two biggest reasons we wrote scale_up.md[*] in the first place were that (1) the notebook would not continue running long enough for the parallel cell to complete at scale (servers were being shut down automatically) and (2) the ZTF function was taking the longest by far so we implemented `multiprocessing` directly in that function, but that could not be used in conjunction with the `multiprocessing` in the notebook's parallel cell. The first might be solved by infrastructure updates and the second by getting ZTF into HATS format -- if so, maybe we can switch `multiprocessing` -> `dask` and run all the calls in parallel with Dask directly from this notebook, at scale.

[*] scale_up.md describes the underlying bash code but doesn't actually execute it at scale -- the user is instructed to do that directly from the command line.

**Review by troyraen — APPROVED** [review]
Thanks for doing this, glad it's working! I left a mixture of comments below ranging from minor naming suggestions to code structure questions that will take a little more work to figure out. Anything you can't or don't want to address here can go into an issue and be assigned to me.

File: light_curves/light_curve_generator.md (hunk 3 above)
Comment (troyraen) [reply light_curves/light_curve_generator.md]: I've seen a new (to me anyway) "git" tab on the LHS of the Fornax console that gives GUI access to changed and staged files, etc. Maybe there's a way to select and commit individual lines/changes from there? I haven't played around with it at all but seems worth checking out.

Comment (bsipocz) [reply light_curves/light_curve_generator.md]: Yeap, I put on the list for my visit to experiment something out.

File: light_curves/code_src/panstarrs_functions.py [hunk 14]
```diff
@@ -244,53 +24,106 @@ def panstarrs_get_lightcurves(sample_table, *, radius=1/3600):
...
+    objectid = matched_df['objectid']
+
+    #make the dataframe of light curves
+    df_lc = pd.DataFrame(
+        dict(flux=pd.to_numeric(flux_panstarrs, errors='coerce').astype(np.float64), 
+             err=pd.to_numeric(err_panstarrs, errors='coerce').astype(np.float64), 
+             time=pd.to_numeric(t_panstarrs, errors='coerce').astype(np.float64), 
+             objectid=pd.to_numeric(objectid, errors='coerce').astype(np.int64), 
+             band=filtername, 
+             label=lab.astype(str))).set_index(["objectid","label", "band", "time"])
```
[the remark this answers is not in the document]
Comment (jkrick) [reply light_curves/code_src/panstarrs_functions.py]: I changed the code to do all the steps within the dataframe assignment to hopefully mitigate this and make it more efficient.  I have not tested it with large samples to know if it is more efficient, but have tested it with small samples to know that it is still functional.

**Review by jkrick — COMMENTED** [review]
I think I have addressed all the comments and made all the requested changes.  Let me know if I missed anything.

**Comment by bsipocz** [comment]
Your comments and experience around working with the dtypes would be a super useful feedback to the lincc team I think.

File: light_curves/code_src/panstarrs_functions.py [hunk 15]
```diff
@@ -244,53 +24,106 @@ def panstarrs_get_lightcurves(sample_table, *, radius=1/3600):
...
+    err_panstarrs = matched_df['psfFluxErr'] *1E3
+    lab = matched_df['label']
+    objectid = matched_df['objectid']
+
+    #make the dataframe of light curves
+    df_lc = pd.DataFrame(
+        dict(flux=pd.to_numeric(flux_panstarrs, errors='coerce').astype(np.float64), 
+             err=pd.to_numeric(err_panstarrs, errors='coerce').astype(np.float64), 
+             time=pd.to_numeric(t_panstarrs, errors='coerce').astype(np.float64), 
+             objectid=pd.to_numeric(objectid, errors='coerce').astype(np.int64),
```
[the remark this answers is not in the document]
Comment (troyraen) [reply light_curves/code_src/panstarrs_functions.py]: Oh interesting. I wonder if it's fundamentally a result of choices made during the catalog import process rather than LSDB itself. I'll try to look at the data types more closely before I import another catalog to try to avoid these hurdles with the ones we put out.

**Comment by troyraen** [comment]
LGTM. Thanks!

File: light_curves/code_src/panstarrs_functions.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (bsipocz) [reply light_curves/code_src/panstarrs_functions.py]: This difference between pyarrow/parquet vs pandas/numpy dtypes certainly came up before, and for end users I see all the reasons to provide the outputs in the latter, even if under the good it keeps mapping to what parquet likes.

[the remark this answers is not in the document]
Comment (troyraen) [reply light_curves/code_src/panstarrs_functions.py]: Good to know, thanks!

**Comment by jkrick** [comment]
> Your comments and experience around working with the dtypes would be a super useful feedback to the lincc team I think.

message received, I put this in as feedback to lsdb in issue #441 over at astronomy-commons/lsdb

**Comment by jkrick** [comment]
@bsipocz when you get a chance, can you please merge this PR?  I don't know why it is failing ci.

**Comment by bsipocz** [comment]
It seems to be something unrelated, but I'm not sure I can dive into the details before I get back home.

**Comment by jkrick** [comment]
I'm not in a hurry, thanks.

**Comment by bsipocz** [comment]
OK, my latest commit should fix the failure. Basically adding some dask would trigger `sktime` to check if `dask[dataframe]` is all installed, and it wasn't.

**Comment by bsipocz** [comment]
Hmm, apparently this is in fact an `sktime` issue: https://github.com/sktime/sktime/issues/7250. Someone reported that downgrading sktime solves the issue, so I was trying that. If it doesn't work, I'll look into more hackery.

**Comment by bsipocz** [comment]
I'm not sure why the classifier notebook run into VM limits now, I'm doing some more debugs and see if it is still OK with GHA. If yes, then I'll bring over the configs from the IRSA notebooks to we can easily opt in GHA build here, too.

**Comment by jkrick** [comment]
> see if it is still OK with GHA
@bsipocz what is "GHA"?

**Comment by bsipocz** [comment]
GHA: GitHub Actions. We use GHA for doing the HTML rendering and hosting the pages with github. But for the PRs we use a different system, and their limitations differ a bit. It was already close to the limit but from below, I suspect adding the new dependencies may changed it worse the worse a bit, but we can't do much about it.

TL;DR, I may need to tweak the configs but I don't think there is much here to do about it. So I would merge this now and figure out issues with circleCI if/when they arise on main.
